lib/ruby_lsp/requests/folding_ranges.rb

```diff
@@ -15,38 +15,206 @@ module Requests
     #   puts "Hello"
     # end # <-- folding range end
     # ```
-    class FoldingRanges < BaseRequest
+    class FoldingRanges < Listener
       extend T::Sig
+      extend T::Generic
 
-      sig { params(document: Document).void }
-      def initialize(document)
-        super
+      ResponseType = type_member { { fixed: T::Array[Interface::FoldingRange] } }
 
-        @ranges = T.let([], T::Array[Interface::FoldingRange])
+      sig { params(comments: T::Array[YARP::Comment], emitter: EventEmitter, queue: Thread::Queue).void }
+      def initialize(comments, emitter, queue)
+        super(emitter, queue)
+
+        @_response = T.let([], ResponseType)
         @requires = T.let([], T::Array[YARP::CallNode])
+        @comments = comments
+
+        emitter.register(
+          self,
+          :on_if,
+          :on_in,
+          :on_rescue,
+          :on_when,
+          :on_interpolated_string,
+          :on_array,
+          :on_block,
+          :on_case,
+          :on_class,
+          :on_module,
+          :on_for,
+          :on_hash,
+          :on_singleton_class,
+          :on_unless,
+          :on_until,
+          :on_while,
+          :on_else,
+          :on_ensure,
+          :on_begin,
+          :on_string_concat,
+          :on_def,
+          :on_call,
+        )
       end
 
-      sig { override.returns(T.all(T::Array[Interface::FoldingRange], Object)) }
-      def run
-        visit(@document.tree)
+      sig { override.returns(ResponseType) }
+      def _response
         push_comment_ranges
         emit_requires_range
-        @ranges
+        @_response
```

Comment:
I remember we wanted `Listener#response` to be idempotent. If that's still the case, we should probably add a flag to guard the above 2 method calls?

Reply:
Good point! I added a guard to make it idempotent.